File: Day11.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

public class Day11 {
    public static void main(String[] args) {
        ArrayList<String> fileData = getFileData("src/InputFile");
        ArrayList<Long> numbers = new ArrayList<Long>();
        HashMap<Long, Long> stoneMap = new HashMap<Long, Long>();
        String line = fileData.get(0);
        String[] stringNumbers = line.split(" ");
        for (String n : stringNumbers) {
            long stoneNumber = Long.parseLong(n);
            numbers.add(Long.parseLong(n));
            long stoneCount = getStoneAmount(stoneMap, stoneNumber);
            stoneMap.put(stoneNumber, stoneCount+1);
        }

        int blinks = 25;
        System.out.println("Part one answer: "+ doBlinks(stoneMap, blinks));
        blinks = 50;
        System.out.println("Part two answer: "+ doBlinks(stoneMap, blinks));

    }

    public static long doBlinks(HashMap<Long, Long> stoneMap, int blinks) {
        for (int i = 0; i < blinks; i++) {
            doSingleBlink(stoneMap);
        }
        long answer = 0;
        for (long value : stoneMap.values()) {
            answer += value;
        }
        return answer;

    }

    public static void doSingleBlink(HashMap<Long, Long> stoneMap) {
        ArrayList<Long> keys = new ArrayList<Long>();
        keys.addAll(stoneMap.keySet());
        HashMap<Long, Long> stoneMapCopy = new HashMap<Long, Long>();
        for (long key : keys) {
            stoneMapCopy.put(key, getStoneAmount(stoneMap, key));
        }
        for (long key : keys) {
            long stoneNumber = getStoneAmount(stoneMapCopy, key);
            String keyString = key + "";

            if (key == 0) {
                stoneMap.put(key+1, getStoneAmount(stoneMap,1) + stoneNumber);
                stoneMap.put(key, getStoneAmount(stoneMap,0) - stoneNumber);
            }

            else if (keyString.length() % 2 == 0) {
                String firstHalf = keyString.substring(0, keyString.length()/2);
                String secondHalf = keyString.substring(keyString.length()/2);
                long stoneLeft = Long.parseLong(firstHalf);
                long stoneRight = Long.parseLong(secondHalf);
                stoneMap.put(stoneLeft, getStoneAmount(stoneMap, stoneLeft) + stoneNumber);
                stoneMap.put(stoneRight, getStoneAmount(stoneMap, stoneRight) + stoneNumber);
                stoneMap.put(key, getStoneAmount(stoneMap, key) - stoneNumber);
            }

            else {
                stoneMap.put(key * 2024, getStoneAmount(stoneMap, key * 2024) + stoneNumber);
                stoneMap.put(key, getStoneAmount(stoneMap, key) - stoneNumber);
            }

        }
    }

    public static ArrayList<String> getFileData(String fileName) {
        ArrayList<String> fileData = new ArrayList<String>();
        try {
            File f = new File(fileName);
            Scanner s = new Scanner(f);
            while (s.hasNextLine()) {
                String line = s.nextLine();
                if (!line.equals(""))
                    fileData.add(line);
            }
            return fileData;
        }
        catch (FileNotFoundException e) {
            return fileData;
        }
    }

    public static long getStoneAmount(HashMap<Long, Long> stones, long key) {
        if (stones.get(key) == null) {
            return (long)0;
        }
        else {
            return stones.get(key);
        }
    }
}

File: Day11.py
from collections import defaultdict

def get_file_data(file_name):
    f = open(file_name)
    data = []
    for line in f:
        data.append(line.rstrip())
    return data

def convert_to_dict(numbers):

    # this will give me a default value of 0 for any stone i don't have
    # idk how to do this in java
    stones = defaultdict(int)
    for stone_number in numbers:
        stone_number = int(stone_number)
        stones[stone_number] += 1

    return stones


def blink_times(blinks, stone_map):
    for i in range(blinks):
        blink(stone_map)
    return

# just keep track of how many stones of each number you have
# since the math isn't complicated, you shouldn't too many of each stone type
# for example:
# if you have 4090 three times, one blink will create 40 three times and 90 three times
# the order of the stones don't really matter here, so a list isn't necessary
def blink(stone_map):
    # have to create a temp dict so I can modify the original dict during iteration
    temporary_stones = dict(stone_map)

    for stone_number, number_of_stones in temporary_stones.items():

        # rule 1
        if stone_number == 0:
            stones[1] += number_of_stones
            stones[0] -= number_of_stones

        # rule 2
        elif len(str(stone_number)) % 2 == 0:
            stone_left = int(str(stone_number)[:int(len(str(stone_number)) / 2)])
            stone_right = int(str(stone_number)[int(len(str(stone_number)) / 2):])

            stones[stone_left] += number_of_stones
            stones[stone_right] += number_of_stones
            stones[stone_number] -= number_of_stones

        # rule 3
        else:
            stones[stone_number * 2024] += number_of_stones
            stones[stone_number] -= number_of_stones
    return


#-----------------------------------------------------------------------------------------

filename = 'InputFile'
#filename = 'sample.txt'
file_data = get_file_data(filename)
numbers = file_data[0].split(" ")

stones = convert_to_dict(numbers)
blink_times(25, stones)
print('Part one answer: ', sum(stones.values()))

stones = convert_to_dict(numbers)
blink_times(75, stones)
print('Part two answer: ', sum(stones.values()))